### PR TITLE
Pre commit checks (WIP)

### PR DESCRIPTION
**Pull request by benjeffery**
#470 
This work-in-progress PR adds pre-commit checks for code quality and style. If anyone has any suggestions for extra checks please comment. Current planned list:
- [x] black
- [x] flake8
- [x] common pre-commit
- [x] isort
- [x] bugbear
- [x] pyupgrade
- [ ] doc checks (both code, text and docstrings)
- [ ] C style
- [ ] integrate with CI

### CODE REVIEW

**Comment by grahamgower** [comment]
Hey @benjeffery. Any plans to start adding type annotations? If so, static type checking could go here too.

**Comment by jeromekelleher** [comment]
> Hey @benjeffery. Any plans to start adding type annotations? If so, static type checking could go here too.

No, we hadn't considered this. I've opened #472 to discuss.

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    -   id: debug-statements
+    -   id: mixed-line-ending
+-   repo: https://github.com/timothycrosley/isort.git
+    rev: 4.3.21-2
+    hooks:
+    -   id: isort
+        files: \.py$
+        # Following https://github.com/timothycrosley/isort/issues/694
+        args: [--multi-line=3, --trailing-comma, --force-grid-wrap=0, --use-parentheses,
+               --line-width=88, --project=tskit,_tskit,msprime]
```
Comment (jeromekelleher) [inline .pre-commit-config.yaml]: msprime is third-party here I think, strictly speaking.

File: python/tests/test_dict_encoding.py [hunk 2]
```diff
@@ -25,10 +25,9 @@
 """
 import unittest
 
+import _tskit as c_module
```
Comment (jeromekelleher) [inline python/tests/test_dict_encoding.py]: This isn't quite right - _tskit is being classified as third party?

File: python/tests/test_genotypes.py [hunk 3]
```diff
@@ -22,18 +22,16 @@
...
-import unittest
 import itertools
 import random
+import unittest
 
-import numpy as np
 import msprime
-
+import numpy as np
+import tests.test_wright_fisher as wf
```
Comment (jeromekelleher) [inline python/tests/test_genotypes.py]: The ``tests`` module probably needs to be registered as local.

**Comment by benjeffery** [comment]
> isort seems to be doing some weird stuff. How's your experience been with it so far?

Not that great - I'm still getting cases where it is disagreeing with `black`. I'm trying out https://github.com/asottile/reorder_python_imports instead.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (benjeffery) [reply .pre-commit-config.yaml]: Yep, I thought so - however I'd seen it grouped with tskit in at one place so was seeing if that looked good throughout.

**Comment by benjeffery** [comment]
`reorder_python_imports` doesn't let you specify `_tskit` as local. Seeing if a quick patch to add the option will do the trick.

**Comment by benjeffery** [comment]
I've filed https://github.com/asottile/reorder_python_imports/pull/123 and will point the pre-commit config at my fork for now.

File: python/tests/test_genotypes.py (hunk 3 above)
Comment (benjeffery) [reply python/tests/test_genotypes.py]: Looks fixed with `reorder_python_imports`

**Comment by jeromekelleher** [comment]
Just wondering if the import sorting/categorising is worth the hassle @benjeffery? Could be a rabbit hole...

**Comment by benjeffery** [comment]
> Just wondering if the import sorting/categorising is worth the hassle @benjeffery? Could be a rabbit hole...

Yeah, I was thinking that too. If the PR I have in now doesn't do the job I'll write it off as it is only a nice to have.

**Comment by benjeffery** [comment]
`bugbear` threw up over a hundred issues which I have fixed - some of these are worth checking as they could be deeper things to fix. For example we had lots of "loop variable not used". For a lot of these it is correct to do so, but worth double checking we're not looping unnecessarily.

File: .flake8 [hunk 4]
```diff
@@ -1,6 +1,7 @@
 [flake8]
 # Based directly on Black's recommendations:
 # https://black.readthedocs.io/en/stable/the_black_code_style.html#line-length
-max-line-length = 80
+max-line-length = 800
```
Comment (jeromekelleher) [inline .flake8]: Seems like a long line! What's this based on?

File: python/tests/test_lowlevel.py [hunk 5]
```diff
@@ -1832,7 +1837,7 @@ def f(mutations):
         u = ts.get_num_nodes()
         for bad_node in [u, u + 1, 2 * u]:
             self.assertRaises(_tskit.LibraryError, f, [(0.1, bad_node)])
-        for bad_pos in [-1, length, length + 1]:
+        for _bad_pos in [-1, length, length + 1]:
```
Comment (jeromekelleher) [inline python/tests/test_lowlevel.py]: This looks like an error, we're not actually testing the bad_pos here .

File: python/tests/test_topology.py [hunk 6]
```diff
@@ -4165,7 +4165,7 @@ def test_overlapping_unary_edges_internal_samples(self):
         trees = [{0: 2}, {0: 2, 1: 2}, {1: 2}]
         for t in ts.trees():
             self.assertEqual(t.parent_dict, trees[t.index])
-        for keep in [True, False]:
+        for _ in [True, False]:
```
Comment (jeromekelleher) [inline python/tests/test_topology.py]: This looks like an error too.

File: python/tests/test_tree_stats.py [hunk 7]
```diff
@@ -2894,7 +2893,7 @@ def naive_branch_allele_frequency_spectrum(
     for j in range(num_windows):
         begin = windows[j]
         end = windows[j + 1]
-        for set_index, sample_set in enumerate(sample_sets):
+        for _set_index, _sample_set in enumerate(sample_sets):
             S = np.zeros(out_dim)
```
Comment (jeromekelleher) [inline python/tests/test_tree_stats.py]: Not sure what's happening here, but looks like an error.

File: python/tests/test_tree_stats.py [hunk 8]
```diff
@@ -3390,7 +3389,7 @@ def f(x):
             return x * (x < n)
 
         # Determine output_dim of the function
-        for mode in ("site", "branch", "node"):
+        for _mode in ("site", "branch", "node"):
             sigma1 = ts.sample_count_stat(sample_sets, f, 3, windows=windows)
```
Comment (jeromekelleher) [inline python/tests/test_tree_stats.py]: Definite error, ``mode`` should be an arg to sample_count_stat

File: python/tskit/drawing.py [hunk 9]
```diff
@@ -26,9 +26,8 @@
 import collections
 import numbers
 
-import svgwrite
 import numpy as np
-
+import svgwrite
```
Comment (jeromekelleher) [inline python/tskit/drawing.py]: svgwrite is third party.

File: python/tskit/stats.py [hunk 10]
```diff
@@ -22,16 +22,15 @@
 """
 Module responsible for computing various statistics on tree sequences.
 """
-import threading
 import struct
 import sys
-
-import numpy as np
+import threading
```
Comment (jeromekelleher) [inline python/tskit/stats.py]: threading is from the stdlib.

**Comment by benjeffery** [comment]
I agree it is getting unwieldy. My approach will be to create a new PR for black+flake8 only, then rebase this PR. That way we don't lose the comments above.

File: python/tests/__init__.py [hunk 11]
```diff
@@ -19,18 +19,16 @@
...
-from .simplify import *  # NOQA
+import tskit
 from . import tsutil
+from .simplify import *  # NOQA
 
-import tskit
+# TODO remove this code and refactor elsewhere.
 
 
-class PythonTree(object):
```
[the remark this answers is not in the document]
Comment (benjeffery) [reply python/tests/__init__.py]: This is from `pyupgrade`. I don't think it can be configured to just skip that one.

File: .flake8 (hunk 4 above)
Comment (benjeffery) [reply .flake8]: That's based on committing debug lines like an idiot. (Although in my defence `git diff master` was massive)

File: python/tests/__init__.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (jeromekelleher) [reply python/tests/__init__.py]: I'm not sure pyupgrade is worth the bother then --- the only thing it's doing that looks useful is to change .format() calls into f strings. But this isn't that big a deal --- nothing wrong with calling .format IMO.

[the remark this answers is not in the document]
Comment (benjeffery) [reply python/tests/__init__.py]: Interesting to find something we disagree on! I find without `(object)` after the basic classes the classes that inherit stand out more. Basically "brackets near classname = you can't understand this class without checking the superclass" is a rule in my mind.

I also find f-strings to be _much_ more readable.

Neither of these things are a religious conviction, but glad we're just picking a standard to end  discussing this stuff in every PR!

[the remark this answers is not in the document]
Comment (jeromekelleher) [reply python/tests/__init__.py]: I'm willing to let go of both of these, neither are strong convictions. Let's keep pyupgrade!

File: python/tests/test_tree_stats.py (hunk 7 above)
Comment (petrelharp) [reply python/tests/test_tree_stats.py]: yep - holdover from a previous version

File: python/tests/tsutil.py [hunk 12]
```diff
@@ -972,7 +972,7 @@ def mean_descendants(ts, reference_sets):
     last_update = np.zeros(ts.num_nodes)
     total_span = np.zeros(ts.num_nodes)
 
-    def update_counts(edge, sign):
+    def update_counts(edge, left, sign):
```
Comment (petrelharp) [inline python/tests/tsutil.py]: wow good catch

**Comment by petrelharp** [comment]
Looking pretty good.

File: python/tests/tsutil.py (hunk 12 above)
Comment (benjeffery) [reply python/tests/tsutil.py]: This was an interesting case - the linter said `left` wasn't being used... but it was! Via the closure and the fact a `for` loop "leaks" its loop variable to the outer scope!

File: .flake8 (hunk 4 above)
Comment (benjeffery) [reply .flake8]: Fixed in the other PR

File: python/tests/test_tree_stats.py (hunk 7 above)
Comment (benjeffery) [reply python/tests/test_tree_stats.py]: Fixed

File: python/tests/test_tree_stats.py (hunk 8 above)
Comment (benjeffery) [reply python/tests/test_tree_stats.py]: Putting `mode` as an arg caused the test to fail. Needed changes to `denom` to match the shape of `sigma3`.

File: python/tskit/drawing.py (hunk 9 above)
Comment (benjeffery) [reply python/tskit/drawing.py]: Fixed

File: python/tskit/stats.py (hunk 10 above)
Comment (benjeffery) [reply python/tskit/stats.py]: Fixed

File: python/tests/test_topology.py (hunk 6 above)
Comment (benjeffery) [reply python/tests/test_topology.py]: Fixed

File: python/tests/test_lowlevel.py (hunk 5 above)
Comment (benjeffery) [reply python/tests/test_lowlevel.py]: Fixed

**Comment by benjeffery** [comment]
>get the basic infrastructure for pre-commit and black in first, and then maybe follow up with other pre-commit checks one-by-one as follow up PRs?
As all the comment threads have resolved, I've split them into their own PRs. Once #475 is merged all the diffs will be clear and show just that tools effect. I'm closing this PR in favour of those.

File: python/tests/test_tree_stats.py (hunk 8 above)
Comment (jeromekelleher) [reply python/tests/test_tree_stats.py]: Good catch!

File: python/tests/tsutil.py (hunk 12 above)
Comment (jeromekelleher) [reply python/tests/tsutil.py]: Wow, good work linter.

**Pull request closed without merging**